Signal/src/util/OWSBackupImportJob.m

```diff
@@ -0,0 +1,506 @@
+{
+    OWSAssert(recordNames.count > 0);
+    OWSAssert(completion);
+
+    DDLogVerbose(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
+
+    // A map of "record name"-to-"downloaded file path".
+    self.downloadedFileMap = [NSMutableDictionary new];
+
+    [self downloadNextFileFromCloud:recordNames completion:completion];
+}
+
+- (void)downloadNextFileFromCloud:(NSMutableArray<NSString *> *)recordNames
+                       completion:(OWSBackupJobCompletion)completion
+{
+    OWSAssert(recordNames);
+    OWSAssert(completion);
+
+    if (self.isComplete) {
+        // Job was aborted.
+        return completion(nil);
+    }
+
+    if (recordNames.count < 1) {
+        // All downloads are complete; exit.
+        return completion(nil);
+    }
+    NSString *recordName = recordNames.lastObject;
+    [recordNames removeLastObject];
+
+    if (![recordName isKindOfClass:[NSString class]]) {
+        DDLogError(@"%@ invalid record name in manifest: %@", self.logTag, [recordName class]);
+        // Invalid record name in the manifest. This may be recoverable.
+        // Ignore this for now and proceed with the other downloads.
+        return [self downloadNextFileFromCloud:recordNames completion:completion];
+    }
+
+    // Use a predictable file path so that multiple "import backup" attempts
+    // will leverage successful file downloads from previous attempts.
+    NSString *tempFilePath = [self.jobTempDirPath stringByAppendingPathComponent:recordName];
+
+    // Skip redundant file download.
+    if ([NSFileManager.defaultManager fileExistsAtPath:tempFilePath]) {
+        [OWSFileSystem protectFileOrFolderAtPath:tempFilePath];
+        self.downloadedFileMap[recordName] = tempFilePath;
+        [self downloadNextFileFromCloud:recordNames completion:completion];
+        return;
+    }
+
+    [OWSBackupAPI downloadFileFromCloudWithRecordName:recordName
+        toFileUrl:[NSURL fileURLWithPath:tempFilePath]
+        success:^{
+            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+                [OWSFileSystem protectFileOrFolderAtPath:tempFilePath];
+                self.downloadedFileMap[recordName] = tempFilePath;
+                [self downloadNextFileFromCloud:recordNames completion:completion];
+            });
+        }
+        failure:^(NSError *error) {
+            completion(error);
+        }];
+}
+
+- (void)restoreAttachmentFiles
+{
+    DDLogVerbose(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
+
+    NSString *attachmentsDirPath = [TSAttachmentStream attachmentsFolder];
+
+    for (NSString *recordName in self.attachmentRecordMap) {
+        if (self.isComplete) {
+            return;
+        }
+
+        NSString *dstRelativePath = self.attachmentRecordMap[recordName];
+        if (!
+            [self restoreFileWithRecordName:recordName dstRelativePath:dstRelativePath dstDirPath:attachmentsDirPath]) {
+            // Attachment-related errors are recoverable and can be ignored.
```

Comment:
We should probably log them though?

Reply:
We do - failures for many of these import/export activities are logged at a lower level, in this case, inside `restoreFileWithRecordName...`.